File: src/lib/agent/wiki-tools.ts
/**
 * Wiki retrieval tools for the chat agent.
 *
 * The shift these implement: instead of dumping the full catalog (every
 * table, every view) into the system prompt and asking the agent to brute-
 * force which is relevant, the system prompt now shows only DOMAINS
 * (4-7 named knowledge areas). The agent picks the relevant domain and uses
 * `browse_wiki` to get its concrete contents — a much tighter retrieval.
 *
 * `search_wiki` is the escape hatch: when the user's question doesn't fit
 * cleanly into a single domain (or no domain is obviously relevant), the
 * agent runs a full-text-ish search across all wiki pages.
 */
import { tool } from "ai";
import { z } from "zod";
import { catalogPool } from "../catalog/db";
import { audit } from "../catalog/queries";

export function wikiTools(opts: { conversationId: string }) {
  return {
    browse_wiki: tool({
      description:
        "Pull a domain's full wiki index — its description, all member tables/views/skills/docs, and any concept pages. Call this FIRST when a question fits a domain you saw in the prompt. Returns enough context to write SQL without inspecting every table.",
      inputSchema: z.object({
        domain_slug: z
          .string()
          .describe("The domain slug from the catalog index (e.g. 'quality-deviations')."),
      }),
      execute: async ({ domain_slug }) => {
        await audit("agent", "tool:browse_wiki", domain_slug, { conversationId: opts.conversationId });
        const dom = await catalogPool.query<{
          id: number;
          slug: string;
          name: string;
          description: string | null;
        }>(
          `SELECT id, slug, name, description FROM wiki_domains WHERE slug = $1`,
          [domain_slug],
        );
        const d = dom.rows[0];
        if (!d) {
          return {
            error: `No domain named "${domain_slug}". Use search_wiki instead if you don't know which domain fits.`,
          };
        }
        const indexPage = await catalogPool.query<{ slug: string; body_md: string }>(
          `SELECT slug, body_md FROM wiki_pages
            WHERE domain_id = $1 AND page_type = 'index'
            ORDER BY updated_at DESC LIMIT 1`,
          [d.id],
        );
        const members = await catalogPool.query<{
          slug: string;
          title: string;
          summary: string | null;
          corpus: string | null;
          page_type: string;
        }>(
          `SELECT slug, title, summary, corpus, page_type
             FROM wiki_pages
            WHERE domain_id = $1 AND page_type IN ('source','concept')
            ORDER BY corpus, title`,
          [d.id],
        );
        return {
          domain: { slug: d.slug, name: d.name, description: d.description },
          index_page_body_md: indexPage.rows[0]?.body_md ?? null,
          members: members.rows,
        };
      },
    }),

    search_wiki: tool({
      description:
        "Search the wiki for pages matching a query. Use this when no single domain obviously fits the user's question, OR when you need to find a specific concept that might be in any domain. Returns up to 10 pages with their slug, title, summary, and the domain they belong to.",
      inputSchema: z.object({
        query: z
          .string()
          .min(2)
          .describe("Search terms. Matches against page title, summary, and body."),
      }),
      execute: async ({ query }) => {
        await audit("agent", "tool:search_wiki", null, { conversationId: opts.conversationId, query: query.slice(0, 200) });
        // Lightweight LIKE search; FTS index is in the plan but not wired yet.
        const pattern = `%${query.replace(/[%_]/g, "")}%`;
        const r = await catalogPool.query<{
          slug: string;
          title: string;
          summary: string | null;
          page_type: string;
          corpus: string | null;
          domain_name: string | null;
          domain_slug: string | null;
        }>(
          `SELECT p.slug, p.title, p.summary, p.page_type, p.corpus,
                  d.name AS domain_name, d.slug AS domain_slug
             FROM wiki_pages p
             LEFT JOIN wiki_domains d ON d.id = p.domain_id
            WHERE p.title ILIKE $1 OR p.summary ILIKE $1 OR p.body_md ILIKE $1
            ORDER BY
              CASE WHEN p.title ILIKE $1 THEN 0
                   WHEN p.summary ILIKE $1 THEN 1
                   ELSE 2
              END,
              p.updated_at DESC
            LIMIT 10`,
          [pattern],
        );
        return { query, hits: r.rows };
      },
    }),

    open_wiki_page: tool({
      description:
        "Open a specific wiki page by its slug (e.g. 'public.deviations' or 'domain/quality-deviations'). Returns the full markdown body, its domain, and pages that link to it. Use this when you need the deep detail of a single source — e.g. to read the columns and common filter patterns before writing SQL.",
      inputSchema: z.object({
        slug: z.string().describe("Full wiki page slug."),
      }),
      execute: async ({ slug }) => {
        await audit("agent", "tool:open_wiki_page", slug, { conversationId: opts.conversationId });
        const r = await catalogPool.query<{
          slug: string;
          title: string;
          summary: string | null;
          body_md: string;
          page_type: string;
          corpus: string | null;
          domain_name: string | null;
          domain_slug: string | null;
        }>(
          `SELECT p.slug, p.title, p.summary, p.body_md, p.page_type, p.corpus,
                  d.name AS domain_name, d.slug AS domain_slug
             FROM wiki_pages p
             LEFT JOIN wiki_domains d ON d.id = p.domain_id
            WHERE p.slug = $1`,
          [slug],
        );
        if (!r.rows[0]) return { error: `No wiki page with slug "${slug}".` };
        const backlinks = await catalogPool.query<{ slug: string; title: string }>(
          `SELECT p.slug, p.title
             FROM wiki_links l JOIN wiki_pages p ON p.id = l.from_page_id
            WHERE l.to_slug = $1
            ORDER BY p.title LIMIT 20`,
          [slug],
        );
        return { page: r.rows[0], backlinks: backlinks.rows };
      },
    }),
  };
}

File: backend/app/agent/wiki_tools.py
"""Wiki retrieval tools for the chat agent.

Port of src/lib/agent/wiki-tools.ts. Surfaces `browse_wiki` (pull a domain's
full index + members), `search_wiki` (LIKE search across all pages), and
`open_wiki_page` (single page by slug, plus backlinks). The agent picks a
domain from the system prompt, then drills in — far tighter retrieval than
dumping every table into context.
"""
from __future__ import annotations

import json
import logging
from typing import Any

from psycopg.rows import dict_row

from ..audit import audit
from ..db import get_conn

log = logging.getLogger(__name__)


TOOLS: list[dict[str, Any]] = [
    {
        "type": "function",
        "function": {
            "name": "browse_wiki",
            "description": (
                "Pull a domain's full wiki index — its description, all member tables/views/skills/docs, "
                "and any concept pages. Call this FIRST when a question fits a domain you saw in the prompt. "
                "Returns enough context to write SQL without inspecting every table."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "domain_slug": {
                        "type": "string",
                        "description": "The domain slug from the catalog index (e.g. 'quality-deviations').",
                    },
                },
                "required": ["domain_slug"],
                "additionalProperties": False,
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "search_wiki",
            "description": (
                "Search the wiki for pages matching a query. Use this when no single domain obviously fits "
                "the user's question, OR when you need to find a specific concept that might be in any "
                "domain. Returns up to 10 pages with their slug, title, summary, and the domain they belong to."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "query": {
                        "type": "string",
                        "minLength": 2,
                        "description": "Search terms. Matches against page title, summary, and body.",
                    },
                },
                "required": ["query"],
                "additionalProperties": False,
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "open_wiki_page",
            "description": (
                "Open a specific wiki page by its slug (e.g. 'public.deviations' or 'domain/quality-deviations'). "
                "Returns the full markdown body, its domain, and pages that link to it. Use this when you need "
                "the deep detail of a single source — e.g. to read the columns and common filter patterns "
                "before writing SQL."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "slug": {"type": "string", "description": "Full wiki page slug."},
                },
                "required": ["slug"],
                "additionalProperties": False,
            },
        },
    },
]


async def run_tool(name: str, args: dict[str, Any], *, conversation_id: str) -> str:
    """Dispatch one wiki tool. Returns marshalled JSON string."""
    try:
        if name == "browse_wiki":
            return _json(await browse_wiki(args["domain_slug"], conversation_id=conversation_id))
        if name == "search_wiki":
            return _json(await search_wiki(args["query"], conversation_id=conversation_id))
        if name == "open_wiki_page":
            return _json(await open_wiki_page(args["slug"], conversation_id=conversation_id))
        return _json({"error": f"Unknown wiki tool: {name}"})
    except Exception as e:  # noqa: BLE001
        log.exception("wiki_tools.run_tool failed: %s", name)
        return _json({"error": f"{type(e).__name__}: {e}"})


async def browse_wiki(domain_slug: str, *, conversation_id: str) -> dict[str, Any]:
    await audit("agent", "tool:browse_wiki", domain_slug, {"conversationId": conversation_id})
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                "SELECT id, slug, name, description FROM wiki_domains WHERE slug = %s",
                (domain_slug,),
            )
            dom = await cur.fetchone()
            if not dom:
                return {
                    "error": (
                        f'No domain named "{domain_slug}". Use search_wiki instead if you '
                        "don't know which domain fits."
                    )
                }
            await cur.execute(
                """
                SELECT slug, body_md FROM wiki_pages
                 WHERE domain_id = %s AND page_type = 'index'
                 ORDER BY updated_at DESC LIMIT 1
                """,
                (dom["id"],),
            )
            index_page = await cur.fetchone()
            await cur.execute(
                """
                SELECT slug, title, summary, corpus, page_type
                  FROM wiki_pages
                 WHERE domain_id = %s AND page_type IN ('source','concept')
                 ORDER BY corpus, title
                """,
                (dom["id"],),
            )
            members = await cur.fetchall()
    return {
        "domain": {"slug": dom["slug"], "name": dom["name"], "description": dom["description"]},
        "index_page_body_md": index_page["body_md"] if index_page else None,
        "members": members,
    }


async def search_wiki(query: str, *, conversation_id: str) -> dict[str, Any]:
    await audit(
        "agent",
        "tool:search_wiki",
        None,
        {"conversationId": conversation_id, "query": query[:200]},
    )
    pattern = "%" + query.replace("%", "").replace("_", "") + "%"
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                SELECT p.slug, p.title, p.summary, p.page_type, p.corpus,
                       d.name AS domain_name, d.slug AS domain_slug
                  FROM wiki_pages p
                  LEFT JOIN wiki_domains d ON d.id = p.domain_id
                 WHERE p.title ILIKE %s OR p.summary ILIKE %s OR p.body_md ILIKE %s
                 ORDER BY
                   CASE WHEN p.title ILIKE %s THEN 0
                        WHEN p.summary ILIKE %s THEN 1
                        ELSE 2
                   END,
                   p.updated_at DESC
                 LIMIT 10
                """,
                (pattern, pattern, pattern, pattern, pattern),
            )
            rows = await cur.fetchall()
    return {"query": query, "hits": rows}


async def open_wiki_page(slug: str, *, conversation_id: str) -> dict[str, Any]:
    await audit("agent", "tool:open_wiki_page", slug, {"conversationId": conversation_id})
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                SELECT p.slug, p.title, p.summary, p.body_md, p.page_type, p.corpus,
                       d.name AS domain_name, d.slug AS domain_slug
                  FROM wiki_pages p
                  LEFT JOIN wiki_domains d ON d.id = p.domain_id
                 WHERE p.slug = %s
                 LIMIT 1
                """,
                (slug,),
            )
            page = await cur.fetchone()
            if not page:
                return {"error": f'No wiki page with slug "{slug}".'}
            await cur.execute(
                """
                SELECT p.slug, p.title
                  FROM wiki_links l
                  JOIN wiki_pages p ON p.id = l.from_page_id
                 WHERE l.to_slug = %s
                 ORDER BY p.title LIMIT 20
                """,
                (slug,),
            )
            backlinks = await cur.fetchall()
    return {"page": page, "backlinks": backlinks}


def _json(value: Any) -> str:
    return json.dumps(value, default=str)
